### DISCUSSION TITLE
Measure space that content uses in a PDF file.

### DISCUSSION THREAD

**Discussion by douglasmiranda** [pymupdf/PyMuPDF#3904] [Q&A]
As a tl;dr; of this essay, I'll just provide with something I say at the end right here in the beggining:

>  More specifically if I could just be able to measure how much of the main body of pdf pages (not including header and footer) is filled with content I would be happy. Like main_body = 100% space of page, content fills 85% of page main_body.

Hi, so what I'm trying to do is, reading a pdf file, (that could be two column), identify those blocks of content and check the space being filled with content inside the body of that file.

Inspired by these articles: 

- https://artifex.com/blog/extracting-text-from-multi-column-pages-a-practical-pymupdf-guide
- https://artifex.com/blog/extract-text-from-a-multi-column-document-using-pymupdf-inpython

Following the examples, I checked I can access the `rect.height`.

```python
import pymupdf
from multi_column import column_boxes

doc = pymupdf.open("pdfs/2-column.pdf")
for page in doc:
    bboxes = column_boxes(page, footer_margin=40, header_margin=40, no_image_text=True)
    for rect in bboxes:
        print(rect.height)
        print(page.get_text(clip=rect, sort=True))

``` 

`multi_column` being: https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/text-extraction/multi_column.py

Which leads me to believe that I can just run through my pages, (ignoring header and footer with the parameters of `column_boxes`), get the content blocks (rect) height (for this example I only care for height). And at the end I'll have how much of the page main body (ignoring header and footer) is filled with textual content. 

Like, lets say I can measure the first block of content in the first image, having 15 centimeters.
 
![2col2tables(1)](https://github.com/user-attachments/assets/9431e23a-ccc6-4c3d-9b95-63b999e79989)

Now for the actual question:

Is that value `rect.height` what I am looking for?

- https://pymupdf.readthedocs.io/en/latest/rect.html#Rect.height

If yes, what type of measure is that? Because in that case I would try to convert that unit of measurement to some other I want.

---

More specifically if I could just be able to measure how much of the main body of pdf pages (not including header and footer) is filled with content I would be happy. Like main_body = 100%  space of page, content fills 85% of page main_body.

**Comment by JorjMcKie** [comment]
All measures are in points, 1 point = 1/72 inch.
Whether or not header/footer is included depends on your margins value.
Note: There is no way to **_compute_** what is a header or footer. So you must specify which parts to ignore via the margins parameter.

**Comment by douglasmiranda** [comment]
Oh I see, so if I know my PDFs are a4 size (595.28 points x  841.89 points).

My real content space is the total a4 size minus margins, footer and header. Then when iterating over the `bboxes`, accessing each `rect` width and height I'll be able to get what I need.

If I come up with some code I'll paste in the comments.

**Comment by douglasmiranda** [comment]
I did a proof of concept, it's actually not hard.

As long all you want is read the text blocks inside a PDF page, and calculate how much % they are occupying of the content area.

So, in my case, all I need to know is how much % of page height the text blocks occupy.

So if I get text blocks `page.get_text("blocks")`  and inspect a single one I have something like (0, 387, 31, 396, 'text xyz', 25, 0), and **396** being the value I need if all I need is how far from the top my text block is. For reference that is y1 if you think in `pymupdf.Rect`.

My full proof of concept is here:

https://gist.github.com/douglasmiranda/5105e1ee71fecf2bdd923e6bef65f95a

Notes:

- it's easy to change this to actually get the whole area (width*height) filled with content. [ref](https://pymupdf.readthedocs.io/en/latest/rect.html#Rect.get_area)
- if your document has a simple formatting of text you might even just need to check the latest text block of each page, no need for a for loop
- if your document has multiple columns, this code may need rework. "may need" because in some 2 column documents you may be dealing with those "justified" columns where in the last page the content is equally divided in 2 columns for the rest of the content.
